### PR TITLE
fix(peers): stop the barrel from requiring optional peer dependencies

### LINKED ISSUE #1735: Barrel import still requires the optional peer date-fns (UniversalFilter), breaking clean-install consumers
## Summary

`src/custom/UniversalFilter.tsx:3` imports `date-fns` at module scope, but `date-fns` is declared an **optional** peer dependency. `UniversalFilter` is reachable from the package barrel, so `import { anything } from '@sistent/sistent'` executes a top-level `require('date-fns')` and throws for any consumer that has not installed it.

This is the same root cause #1732 fixed for `@mui/x-date-pickers`. That PR resolved one instance and revealed this one behind it.

## Reproduction (against published 0.21.39)

```bash
mkdir /tmp/repro && cd /tmp/repro && npm init -y
npm i @sistent/sistent@0.21.39 react react-dom @mui/material @mui/system @emotion/react @emotion/styled
node -e "require('@sistent/sistent')"
```

```
Error: Cannot find module 'date-fns'
Require stack:
- /tmp/repro/node_modules/@sistent/sistent/dist/index.js
```

Installing `date-fns` makes it resolve cleanly, confirming it is the only remaining instance.

Before #1732 the same reproduction failed on `Cannot find module '@mui/x-date-pickers/AdapterDateFns'`, so this is **not a regression** - it is the next optional peer in line.

## Impact

`layer5labs/meshery-extensions` (`ui/package.json`) depends on `@sistent/sistent` and has **neither** `date-fns` nor `@mui/x-date-pickers`. Any test or module there that imports anything from the barrel still fails on a clean install, with a message that names sistent rather than the missing peer - the diagnostic problem #1732 set out to remove.

`meshery/meshery` (`ui/package.json`) has both (`date-fns@^4.4.0`, `@mui/x-date-pickers@^9.6.0`) and is unaffected.

## Root cause

The declaration and the usage disagree. `date-fns` is marked optional:

```jsonc
"peerDependenciesMeta": { "date-fns": { "optional": true } }
```

but `UniversalFilter` uses it **unconditionally**. An optional dependency has to be either used conditionally or declared honestly - right now it is neither.

## Scope

Exactly one instance remains. `@mui/x-date-pickers` is fully resolved (only type-only imports, erased at runtime, plus dynamic `import()`):

| File | Import | Runtime require? |
|---|---|---|
| `src/custom/UniversalFilter.tsx:3` | `import { subDays, subMonths, subYears } from 'date-fns'` | **yes - the bug** |
| `src/base/DateTimePicker/DateTimePicker.tsx:1` | `import type ...` | no |
| `src/base/DateTimePicker/index.tsx:1` | `import type ...` | no |
| `src/base/DateTimePicker/DateTimePicker.tsx:21-23` | dynamic `import()` | no |

## Options

**A. Declare `date-fns` a real dependency.** One line in `package.json`; makes the declaration match the unconditional usage. Adds a dependency for every consumer.

**B. Drop `date-fns` here in favour of `moment`,** which is already a hard `dependency` of sistent (`^2.30.1`), so it is always installed. The usage is three calls inside lazily-invoked `getRange` closures:

```ts
{ label: 'Last 7 days',   getRange: () => ({ startDate: subDays(new Date(), 7),    endDate: new Date() }) },
{ label: 'Last 3 months', getRange: () => ({ startDate: subMonths(new Date(), 3),  endDate: new Date() }) },
{ label: 'Last 1 year',   getRange: () => ({ startDate: subYears(new Date(), 1),   endDate: new Date() }) },
```

`moment().subtract(n, 'days'|'months'|'years').toDate()` is equivalent for these, including the end-of-month clamping `subMonths` does. Zero new dependencies, and it removes an optional peer from the barrel's runtime graph entirely.

B is preferable on dependency hygiene. Flagging that `UniversalFilter.tsx` may have work in flight, so B is best folded into whoever is already touching it rather than landed in parallel.

## Suggested guard

Whichever option is taken, the failure mode is invisible to this repo's CI because the repo always has its own devDependencies installed. A check that imports the built barrel with only the **required** peers present would catch the next instance:

```bash
node -e "require('./dist/index.js')"   # in a dir without the optional peers
```

### PR DESCRIPTION
**Pull request by marblom007**
Fixes #1735. Completes the optional-peer work started in #1732.

## The bug class

`@mui/x-date-pickers` and `date-fns` are both declared **optional** under `peerDependenciesMeta`, so a consumer is entitled to skip them. But `src/index.tsx` re-exports nearly everything, which means a module-scope import of an optional peer in *any* barrel-reachable file breaks that promise for everyone - not just for people who render the component that needs it:

```js
import { Button } from '@sistent/sistent';
// Error: Cannot find module 'date-fns'
```

The message names sistent rather than the peer the consumer deliberately did not install, which is what makes it expensive to diagnose downstream.

#1732 fixed one instance by deferring `@mui/x-date-pickers` to `React.lazy`. **One instance remained:** `UniversalFilter` still imported `subDays` / `subMonths` / `subYears` from `date-fns` at module scope for its quick-range presets, so the barrel was still unusable on a clean install. Meshery UI happens to have `date-fns` installed and never saw it; Kanvas does not, and is blocked on this (layer5labs/meshery-extensions#4303).

## The fix

New `src/utils/date.utils.ts` - `subtractDays` / `subtractMonths` / `subtractYears`, dependency-free.

Deliberately a leaf module with no imports at all, and deliberately **not** folded into `time.utils.tsx`, which pulls in `moment` and a React component: pure date arithmetic should be reachable without dragging a rendering stack behind it. `moment` was considered and rejected for the same reason - swapping one heavyweight import for another does not fix an import-graph problem.

Semantics match `date-fns` exactly:
- end-of-month clamping - `subtractMonths(May 31, 3)` is Feb 28, not Mar 3 (plain `setMonth` overflows)
- `subtractYears(Feb 29 2024, 1)` is Feb 28 2023, not Mar 1
- local-time day arithmetic rather than `n * 86400000` ms, so a range spanning a DST boundary keeps its wall-clock time instead of drifting an hour

## Preventing recurrence

`src/__testing__/optionalPeerDependencies.test.ts` guards the **whole class**, not this one component. It walks the source tree and fails on any eager import of an optional peer, listing the offending paths - so the failure message is also the remediation. It additionally asserts the peers are still marked optional, so if that premise ever changes the assertions get revisited rather than silently kept.

This matters because, as noted in #1736, **CI cannot otherwise catch this**: the repo always has its own devDependencies installed, so the breakage only appears in a downstream clean install. This test makes it visible in-repo.

`src/__testing__/date.utils.test.ts` covers the helpers themselves, including the clamping and DST cases above.

## Also carried in this change

- **`DateTimePicker`** wraps its lazy peer import so a missing peer surfaces as a message naming the peer, rather than a raw module-not-found thrown mid-render from inside a promise. Deferring the resolution deferred the failure too, and an unactionable async crash is no improvement on an unactionable sync one.
- **`time.utils.tsx`** imports `CustomTooltip` from its leaf path instead of the `../custom` barrel. That closed a `custom/X -> utils -> custom/index -> X` cycle and forced every consumer of a date helper to load the entire component library (including, via `Markdown`, ESM-only `react-markdown`).
- **`permissions.tsx`** types `createCanShow`'s bus as a structural `ReasonEventPublisher`. `EventBus<T>` is invariant in `T` - it both accepts `T` in `publish` and yields `T` from `on` - so the host's `EventBus<MesheryExtensionEvent>`, the bus the contract tells every host to declare, was **not** assignable to `EventBus<ReasonEvent>` and the integration failed to compile. It also publishes via the named `MESHERY_EXTENSION_EVENT.MissingPermission` handle so a contract rename breaks this publish site rather than silently ceasing to match the subscriber across the bundle boundary, and drops a stray `console.log` from the click path.
- **Barrel** exports `HasKeyProps`, `InvertAction`, `ReasonEvent` and `ReasonEventPublisher` alongside `createCanShow`; without them a consumer cannot name the types of the wrapper it builds around the returned component and falls straight back to `any`.

## Verification

- `npm test` - 24 suites, 443 tests passing, including the 32 new ones
- `npm run lint` clean; `prettier --check` clean on every touched file (the ~82 pre-existing failures elsewhere are unrelated, see #1736)
- `npm run build` clean
- **By content, on the built output:** the string `date-fns` no longer appears as a module specifier anywhere in `dist/index.js` or `dist/index.mjs`. The three `@mui/x-date-pickers/*` specifiers that remain are inside `import(...)` calls only - no static import survives.
- **Differential against the real library:** 167,637 comparisons of `subtractDays` / `subtractMonths` / `subtractYears` against `date-fns@4.4.0` - every day from 2019-2027 at three times of day across a range of amounts - produced **0 mismatches**. The "semantics match exactly" claim is measured, not asserted.
- **End to end:** with `node_modules/date-fns` and `node_modules/@mui/x-date-pickers` both removed, `require('./dist/index.js')` succeeds and resolves 719 exports. That is the exact scenario that throws on `master` today.

## Guard hardening (from review)

The first version of the guard only matched `import ... from 'x'` and `require('x')`. Review caught that it missed several forms that resolve the module just as eagerly, and one of them mattered a great deal:

- **`export ... from 'x'`** - `src/index.tsx` is built almost entirely out of `export ... from`. The barrel is precisely where this bug class originates, and the guard was blind to the barrel's own syntax. Adding `export { subDays } from 'date-fns'` to `src/index.tsx` passed the original guard; it now fails and names `index.tsx`.
- **`import 'x'`** - a side-effect import binds nothing, which is exactly why it is easy to miss by eye.
- the subpath variants of both.

The detector is now pinned by its own cases (13 assertions for what it must and must not flag: dynamic `import()`, type-only imports and re-exports, unrelated specifiers, prose mentions). A scanner that quietly stops matching a form reports the same clean result as a clean codebase, so the pattern needed to stop being the untested part of the test.

`import type` stripping was widened to `export type` to match - otherwise the new `export ... from` branch would fail a correct type re-export, and the only way to satisfy the guard would be to delete a type.

Separately, `react` and `react-dom` are also marked optional in `package.json` but are not scanned (every component imports React at module scope and always will). They are now named in an explicit exemption list, and a new assertion requires the scanned set plus the exemptions to equal the declared optional peers exactly - so **marking a new peer optional fails the suite** rather than silently creating a peer nothing guards. Verified by mutation with `@rjsf/core`.

Also fixed: the `DateTimePicker` remediation message named a `<UniversalFilter dateRange>` prop that does not exist (it is `datePicker`). An error message whose entire purpose is to be actionable has to name the real API.

## Out of scope, but noticed

`dist/index.mjs` emits `import ... from "lodash/debounce"` (from `src/custom/SearchBar.tsx`), which Node's ESM resolver rejects - it needs the `.js` extension. Bundlers paper over it, so it does not affect Meshery UI or Kanvas, and it is pre-existing on `master` and unrelated to optional peers. Filed separately rather than folded in here.

## Note for reviewers

`master` picked up two contract commits (5eeff36, 61619b2) after the base this work was written against, which refined `mesheryExtensionContract.ts` and its tests along a different line. Those landed versions are kept as-is here - this PR does not touch either file, so nothing already on `master` is reverted.

#1736 lists #1735 as a "known live instance" of the barrel trap; that line goes stale once this merges and is worth a one-line update there.

## Summary by CodeRabbit

- **New Features**
  - Added dependency-free date helpers (`subtractDays`, `subtractMonths`, `subtractYears`) that return new `Date` values, preserve time-of-day, and clamp to valid month-end/leap-year dates.
  - Updated quick date-range filters and exported the new date utilities.
- **Bug Fixes**
  - Improved optional date-picker peer dependency handling with clearer, deferred load errors.
  - Standardized missing-permission event typing and re-exported permission types from the main entrypoint.
- **Tests**
  - Added Jest coverage for date subtraction edge cases and enforced optional peer dependency import safety.

### CODE REVIEW

File: src/__testing__/optionalPeerDependencies.test.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+/** `import ... from 'x'` / `require('x')`, but NOT `await import('x')`. */
+const eagerImportPattern = (specifier: string) =>
+  new RegExp(
+    String.raw`(?:^|[^.\w])(?:import\s[^;]*?from\s*|require\s*\()\s*['"]${specifier.replace(
+      /[/\\^$*+?.()|[\]{}]/g,
+      '\\$&'
+    )}(?:/[^'"]*)?['"]`,
+    'm'
+  );
```
[the remark this answers is not in the document]
Comment (marblom007) [reply src/__testing__/optionalPeerDependencies.test.ts]: Correct, and the re-export case was the important one: `src/index.tsx` is built almost entirely out of `export ... from`, so the guard was blind to the exact syntax of the file that turns one module's import into every consumer's problem. The pattern now matches `export ... from 'x'`, the side-effect `import 'x'`, and the subpath variants of both, while still ignoring `await import('x')`. Verified by mutation: adding `export { subDays } from 'date-fns'` to the barrel now fails the suite and names `index.tsx`, where before it passed. The detector is also pinned by its own cases now (13 assertions for what it must and must not flag), since a scanner that quietly stops matching reports the same clean result as a clean codebase. `import type` stripping was widened to `export type` so the new branch does not fail a correct type re-export. On the ast-grep ReDoS note: not applicable here - the inputs are this repo's own source files, and the only unbounded quantifier is `[^;]*?` which is bounded by a statement terminator.

File: src/base/DateTimePicker/DateTimePicker.tsx [hunk 2]
```diff
@@ -15,13 +15,26 @@ import React from 'react';
...
+      // Deferring the resolution also defers the failure, so the raw
+      // module-not-found surfaces mid-render from inside a promise, far from
+      // any import statement. Naming the peer here is the whole reason the
+      // eager import was worth trading away — an unactionable async crash is
+      // not an improvement on an unactionable sync one.
+      throw new Error(
+        `<DateTimePicker> requires the optional peer dependencies ${OPTIONAL_PEERS}. ` +
+          `Install them, or do not render <DateTimePicker> / <UniversalFilter dateRange>.`,
+        { cause }
+      );
```
[the remark this answers is not in the document]
Comment (marblom007) [reply src/base/DateTimePicker/DateTimePicker.tsx:36]: Confirmed and fixed - the prop is `datePicker`, not `dateRange` (`UniversalFilterProps` declares `datePicker?: boolean`). Naming a nonexistent prop in the one message whose entire purpose is to tell the consumer what to do was worse than saying nothing, so this was worth catching.

**Comment by marblom007** [comment]
@coderabbitai review

**Comment by marblom007** [comment]
@coderabbitai review
